### PR TITLE
Better evaluate impact in memory of range selection

### PR DESCRIPTION
**Pull request by roidelapluie**
Signed-off-by: Julien Pivotto <roidelapluie@inuits.eu>

### CODE REVIEW

File: promql/engine.go [hunk 1]
```diff
@@ -1158,10 +1158,11 @@ func (ev *evaluator) eval(expr parser.Expr) parser.Value {
 				// Only buffer stepRange milliseconds from the second step on.
 				it.ReduceDelta(stepRange)
 			}
+			points = points[:0]
```
Comment (brian-brazil) [inline promql/engine.go:1161]: It's a bit weird to do this after rather than before, it makes the logic less obvious.

**Comment by roidelapluie** [comment]
Yes I am working on test now and we are somehow still overcounting.

**Comment by brian-brazil** [comment]
This PR only corrects the overcount after an entire series is processed, not on each step.

File: promql/engine.go [hunk 2]
```diff
@@ -1144,6 +1145,7 @@ func (ev *evaluator) eval(expr parser.Expr) parser.Value {
 				mint := maxt - selRange
 				// Evaluate the matrix selector for this series for this step.
 				points = ev.matrixIterSlice(it, mint, maxt, points)
+				ev.currentSamples = tmpEvalPoints
```
Comment (brian-brazil) [inline promql/engine.go:1148]: Not quite, the points are still in memory.

We probably want to rethink how matrixIterSlice tracking works - maybe make it external to matrixIterSlice?

**Pull request closed without merging**